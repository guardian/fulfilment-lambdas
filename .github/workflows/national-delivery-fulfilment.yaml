name: "CI: National Delivery Fulfilment"
on: [push]
jobs:
  ci:
    permissions:
      id-token: write
      contents: read
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v3
      - name: Setup node
        uses: actions/setup-node@v3
        with:
          node-version-file: national-delivery-fulfilment/.nvmrc
          cache: "yarn"
      - run: |
          yarn install
          yarn build
          yarn package
        working-directory: national-delivery-fulfilment
      - name: CDK synth
        run: |
          yarn install
          yarn tsc
          yarn lint
          yarn test
          yarn synth
        working-directory: national-delivery-fulfilment/cdk
      - name: AWS Auth
        uses: aws-actions/configure-aws-credentials@v2
        with:
          role-to-assume: ${{ secrets.GU_RIFF_RAFF_ROLE_ARN }}
          aws-region: eu-west-1
      - name: Upload to riff-raff
        uses: guardian/actions-riff-raff@v2
        with:
          configPath: national-delivery-fulfilment/riff-raff.yaml
          projectName: MemSub::Fulfilment::NationalDelivery
          contentDirectories: |
            cloudformation:
              - national-delivery-fulfilment/cdk/cdk.out/NationalDeliveryFulfilment-CODE.template.json
              - national-delivery-fulfilment/cdk/cdk.out/NationalDeliveryFulfilment-PROD.template.json
            national-delivery-fulfilment:
              - national-delivery-fulfilment/national-delivery-fulfilment.zip
