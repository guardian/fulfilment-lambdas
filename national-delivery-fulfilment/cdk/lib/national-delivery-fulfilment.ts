//import { join } from 'path';
import { GuScheduledLambda } from '@guardian/cdk';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack } from '@guardian/cdk/lib/constructs/core';
import {GuAllowPolicy} from "@guardian/cdk/lib/constructs/iam";
import type { App } from 'aws-cdk-lib';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { BlockPublicAccess, Bucket } from 'aws-cdk-lib/aws-s3';
//import { CfnInclude } from 'aws-cdk-lib/cloudformation-include';


export class NationalDeliveryFulfilment extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'national-delivery-fulfilment';
		const nationalDeliveryFulfilmentLambda = new GuScheduledLambda(
			this,
			'national-delivery-fulfilment-lambda',
			{
				description: 'A lambda to handle fulfilment for national delivery',
				functionName: `membership-${app}-${this.stage}`,
				handler: 'national-delivery-fulfilment/index.handler',
				runtime: Runtime.NODEJS_18_X,
				memorySize: 1024,
				fileName: `${app}.zip`,
				app: app,
				rules: [],
				monitoringConfiguration: { noMonitoring: true },
			},
		);

		const bucketName = `gu-national-delivery-fulfilment-${this.stage.toLowerCase()}`

        const dataBucket = new Bucket(this, 'DataBucket', {
          bucketName: bucketName,
        });

        nationalDeliveryFulfilmentLambda.addToRolePolicy(
          new PolicyStatement({
                actions: ['s3:PutObject', "s3:PutObjectAcl"],
              	effect: Effect.ALLOW,
              	resources: [`arn:aws:s3:::${bucketName}/*`],
          }),
        );
	}
}
