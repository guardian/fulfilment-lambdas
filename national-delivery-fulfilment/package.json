{
  "name": "national-delivery-fulfilment",
  "version": "1.0.0",
  "description": "National Delivery fulfilment",
  "main": "index.js",
  "license": "MIT",
  "private": true,
  "dependencies": {
    "@aws-sdk/client-s3": "^3.391.0",
    "@types/aws-lambda": "^8.10.119",
    "@types/node": "^20.5.0",
    "aws-lambda": "^1.0.7",
    "aws-sdk": "^2.1439.0"
  },
  "devDependencies": {
    "esbuild": "^0.19.0"
  },
  "scripts": {
    "build": "yarn esbuild --bundle --platform=node --target=node18 --outfile=dist/index.js src/index.ts",
    "package": "cd dist; zip -qr ../national-delivery-fulfilment.zip ./*"
  }
}
