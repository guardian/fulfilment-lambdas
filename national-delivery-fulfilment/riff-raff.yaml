regions:
  - eu-west-1
stacks:
  - membership
deployments:
  cloudformation:
    type: cloud-formation
    app: national-delivery-fulfilment
    parameters:
      templateStagePaths:
        CODE: NationalDeliveryFulfilment-CODE.template.json
        PROD: NationalDeliveryFulfilment-PROD.template.json
  national-delivery-fulfilment:
    type: aws-lambda
    parameters:
      bucketSsmLookup: true
      fileName: national-delivery-fulfilment.zip
      functionNames:
        - -national-delivery-fulfilment-
    dependencies: [cloudformation]
