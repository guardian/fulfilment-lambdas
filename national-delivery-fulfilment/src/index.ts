import { APIGatewayProxyCallback, Context } from "aws-lambda";
import { main } from "./main";

export async function handler(
  event: APIGatewayEvent,
  context: Context,
  callback: APIGatewayProxyCallback
) {
  await main();
}

export interface APIGatewayEvent {
  headers: Record<string, string | undefined>;
  path: string;
  body: string;
}
