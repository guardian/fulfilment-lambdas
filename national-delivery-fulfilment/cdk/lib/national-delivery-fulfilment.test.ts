import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { NationalDeliveryFulfilment } from './national-delivery-fulfilment';

describe('The NationalDeliveryFulfilment stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const codeStack = new NationalDeliveryFulfilment(
			app,
			'NationalDeliveryFulfilment-CODE',
			{ stack: 'membership', stage: 'CODE' },
		);
		const prodStack = new NationalDeliveryFulfilment(
			app,
			'NationalDeliveryFulfilment-PROD',
			{ stack: 'membership', stage: 'PROD' },
		);
		expect(Template.fromStack(codeStack).toJSON()).toMatchSnapshot();
		expect(Template.fromStack(prodStack).toJSON()).toMatchSnapshot();
	});
});
