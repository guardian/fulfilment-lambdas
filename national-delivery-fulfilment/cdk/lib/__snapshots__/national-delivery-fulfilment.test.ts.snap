// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The NationalDeliveryFulfilment stack matches the snapshot 1`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "nationaldeliveryfulfilmentlambda2870F422": {
      "DependsOn": [
        "nationaldeliveryfulfilmentlambdaServiceRoleDefaultPolicy8EA432E5",
        "nationaldeliveryfulfilmentlambdaServiceRole616D4F71",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/CODE/national-delivery-fulfilment/national-delivery-fulfilment.zip",
        },
        "Description": "A lambda to handle fulfilment for national delivery",
        "Environment": {
          "Variables": {
            "APP": "national-delivery-fulfilment",
            "STACK": "membership",
            "STAGE": "CODE",
          },
        },
        "FunctionName": "membership-national-delivery-fulfilment-CODE",
        "Handler": "national-delivery-fulfilment/lambda/lambda.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "nationaldeliveryfulfilmentlambdaServiceRole616D4F71",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "national-delivery-fulfilment",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/fulfilment-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "nationaldeliveryfulfilmentlambdaServiceRole616D4F71": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "national-delivery-fulfilment",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/fulfilment-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "CODE",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "nationaldeliveryfulfilmentlambdaServiceRoleDefaultPolicy8EA432E5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/CODE/national-delivery-fulfilment/national-delivery-fulfilment.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/national-delivery-fulfilment",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/CODE/membership/national-delivery-fulfilment/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "nationaldeliveryfulfilmentlambdaServiceRoleDefaultPolicy8EA432E5",
        "Roles": [
          {
            "Ref": "nationaldeliveryfulfilmentlambdaServiceRole616D4F71",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;

exports[`The NationalDeliveryFulfilment stack matches the snapshot 2`] = `
{
  "Metadata": {
    "gu:cdk:constructs": [
      "GuDistributionBucketParameter",
      "GuScheduledLambda",
    ],
    "gu:cdk:version": "TEST",
  },
  "Parameters": {
    "DistributionBucketName": {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "nationaldeliveryfulfilmentlambda2870F422": {
      "DependsOn": [
        "nationaldeliveryfulfilmentlambdaServiceRoleDefaultPolicy8EA432E5",
        "nationaldeliveryfulfilmentlambdaServiceRole616D4F71",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Ref": "DistributionBucketName",
          },
          "S3Key": "membership/PROD/national-delivery-fulfilment/national-delivery-fulfilment.zip",
        },
        "Description": "A lambda to handle fulfilment for national delivery",
        "Environment": {
          "Variables": {
            "APP": "national-delivery-fulfilment",
            "STACK": "membership",
            "STAGE": "PROD",
          },
        },
        "FunctionName": "membership-national-delivery-fulfilment-PROD",
        "Handler": "national-delivery-fulfilment/lambda/lambda.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "nationaldeliveryfulfilmentlambdaServiceRole616D4F71",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Tags": [
          {
            "Key": "App",
            "Value": "national-delivery-fulfilment",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/fulfilment-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
        "Timeout": 30,
      },
      "Type": "AWS::Lambda::Function",
    },
    "nationaldeliveryfulfilmentlambdaServiceRole616D4F71": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "App",
            "Value": "national-delivery-fulfilment",
          },
          {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          {
            "Key": "gu:repo",
            "Value": "guardian/fulfilment-lambdas",
          },
          {
            "Key": "Stack",
            "Value": "membership",
          },
          {
            "Key": "Stage",
            "Value": "PROD",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "nationaldeliveryfulfilmentlambdaServiceRoleDefaultPolicy8EA432E5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      {
                        "Ref": "DistributionBucketName",
                      },
                      "/membership/PROD/national-delivery-fulfilment/national-delivery-fulfilment.zip",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/national-delivery-fulfilment",
                  ],
                ],
              },
            },
            {
              "Action": [
                "ssm:GetParameters",
                "ssm:GetParameter",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:aws:ssm:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/PROD/membership/national-delivery-fulfilment/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "nationaldeliveryfulfilmentlambdaServiceRoleDefaultPolicy8EA432E5",
        "Roles": [
          {
            "Ref": "nationaldeliveryfulfilmentlambdaServiceRole616D4F71",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
